File: 13 Roman to Integer/solution.kt
class Solution {
    fun romanToInt(s: String): Int {
        var r = 0
        var i = 0 
        while (i < s.length) {
            val c = s[i++] 
            
            when (c) {
                'I' -> {
                    if (i < s.length && s [i] == 'V') {
                        i++ 
                        r += 4 
                    } else if (i < s.length && s [i] == 'X') {
                        i++ 
                        r += 9 
                    } else {
                        r++ 
                        
                        while (i < s.length && s [i] == 'I') {
                            r++ 
                            i++ 
                        }
                    }
                     
                }
                'V' -> {
                    r += 5 
                     
                }
                'X' -> {
                    if (i < s.length && s [i] == 'L') {
                        i++ 
                        r += 40 
                    } else if (i < s.length && s [i] == 'C') {
                        i++ 
                        r += 90 
                    } else {
                        r += 10 
                        
                        while (i < s.length && s [i] == 'X') {
                            r += 10 
                            i++ 
                        }
                    }
                     
                }
                'L' -> {
                    r += 50 
                     
                }
                'C' -> {
                    if (i < s.length && s [i] == 'D') {
                        i++ 
                        r += 400 
                    } else if (i < s.length && s [i] == 'M') {
                        i++ 
                        r += 900 
                    } else {
                        r += 100 
                        
                        while (i < s.length && s [i] == 'C') {
                            r += 100 
                            i++ 
                        }
                    }
                     
                }
                'D' -> {
                    r += 500 
                     
                }
                'M' -> {
                    r += 1000 
                     
                }
                else -> {}
            }
        }
        return r 
    }
}

File: 13 Roman to Integer/solution.ex
defmodule Solution do
  @two %{"IV" => 4, "IX" => 9, "XL" => 40, "XC" => 90, "CD" => 400, "CM" => 900}
  @one %{"I" => 1, "V" => 5, "X" => 10, "L" => 50, "C" => 100, "D" => 500, "M" => 1000}

  @spec roman_to_int(s :: String.t) :: integer
  def roman_to_int(s) do
    len = String.length s
    case len do
        _ when len > 1 ->
            case @two[String.slice(s, 0..1)] do
                nil -> roman_to_int(String.slice(s, 1..-1)) + @one[String.slice(s, 0..0)]
                val -> roman_to_int(String.slice(s, 2..-1)) + val
            end
        _ when len > 0 -> roman_to_int(String.slice(s, 1..-1)) + @one[String.slice(s, 0..0)]
        _ -> 0
    end
  end
end